Tools/ardupilotwaf/toolchain.py

```diff
@@ -1,93 +1,120 @@
+    def new_configure(cfg):
+        if cfg.env.TOOLCHAIN == 'native':
+            original_configure(cfg)
+            return
+
+        last_exception = None
+        for name in compiler_names:
+            cfg.env.stash()
+            try:
+                cfg.env[var] = '%s-%s' % (cfg.env.TOOLCHAIN, name)
+                original_configure(cfg)
+            except Errors.ConfigurationError as e:
+                cfg.env.revert()
+                last_exception = e
+            else:
+                cfg.env.commit()
+                return
+        raise last_exception
+    tool_module.configure = new_configure
+
+_set_toolchain_prefix_wrapper(gxx, 'CXX', ('g++', 'c++'))
+_set_toolchain_prefix_wrapper(gcc, 'CC', ('gcc', 'cc'))
+
+def _clang_cross_support(cfg):
+    if _clang_cross_support.called:
+        return
+
+    prefix = cfg.env.TOOLCHAIN + '-'
+
+    cfg.find_program(prefix + 'gcc', var='CROSS_GCC')
+
+    environ = dict(os.environ)
+    if 'TOOLCHAIN_CROSS_AR' in environ:
+        # avoid OS's environment to mess up toolchain path finding
+        del environ['TOOLCHAIN_CROSS_AR']
+    try:
+        cfg.find_program(
+            prefix + 'ar',
+            var='TOOLCHAIN_CROSS_AR',
+            environ=environ,
+        )
+    except Errors.ConfigurationError as e:
+        cfg.fatal('toolchain: clang: couldn\'t find toolchain path', ex=e)
+
+    toolchain_path = os.path.join(cfg.env.TOOLCHAIN_CROSS_AR[0], '..', '..')
+    toolchain_path = os.path.abspath(toolchain_path)
+    cfg.msg('Using toolchain path for clang', toolchain_path)
+
+    sysroot = cfg.cmd_and_log(
+        [cfg.env.CROSS_GCC[0], '--print-sysroot'],
+        quiet=Context.BOTH,
+    ).strip()
+
+    cfg.env.CLANG_FLAGS = [
+        '--target=' + cfg.env.TOOLCHAIN,
+        '--gcc-toolchain=' + toolchain_path,
+        '--sysroot=' + sysroot,
+        '-B' + os.path.join(toolchain_path, 'bin')
+    ]
+
+_clang_cross_support.called = False
+
+def _set_clang_crosscompilation_wrapper(tool_module):
+    original_configure = tool_module.configure
+    def new_configure(cfg):
+        if cfg.env.TOOLCHAIN == 'native':
+            original_configure(cfg)
+            return
+
+        cfg.env.stash()
+        try:
+            _clang_cross_support(cfg)
+            original_configure(cfg)
+        except Errors.ConfigurationError as e:
+            cfg.env.revert()
+            raise
+        else:
+            cfg.env.commit()
+    tool_module.configure = new_configure
+
+_set_clang_crosscompilation_wrapper(clang)
+_set_clang_crosscompilation_wrapper(clangxx)
 
-                if filename not in os.path.basename(realpath):
-                    continue
-                else:
-                    return os.path.dirname(realpath)
+def configure(cfg):
+    if cfg.env.TOOLCHAIN == 'native':
+        cfg.load('compiler_cxx compiler_c')
+        return
 
-            else:
-                return os.path.dirname(path)
+    cfg.env.AR = cfg.env.TOOLCHAIN + '-ar'
```

Comment:
This is forcing AR to always be in the format _prefix_-ar, it should only be set for GCC or Clang.